File: src/utils/path-safety.js
// Mercury Code - Shared Path Safety Utilities
// Unified workspace boundary checks, symlink-safe resolution, and git ref validation.
// Used by: tools/executor.js, sandbox.js, repl.js, rollback.js
//
// All functions handle Windows path normalization (forward/back slashes, case-insensitive).

import path from "node:path";
import fs from "node:fs";

/**
 * Normalize a path for comparison: resolve to absolute, normalize separators,
 * and on Windows lowercase the entire path for case-insensitive comparison.
 * @param {string} p
 * @returns {string}
 */
function _normForCompare(p) {
  let resolved = path.resolve(p);
  // Normalize separators (Windows backslashes -> forward slashes -> native)
  resolved = path.normalize(resolved);
  // Case-insensitive on Windows
  if (process.platform === "win32") {
    resolved = resolved.toLowerCase();
  }
  return resolved;
}

/**
 * Unified workspace boundary check (symlink-safe).
 * Uses fs.realpathSync to resolve symlinks before comparison.
 * Falls back to path.resolve if realpath fails (e.g. workspace doesn't exist).
 * Supports Windows case-insensitive comparison and both slash types.
 *
 * @param {string} filePath  - Path to check
 * @param {string} workspace - Workspace root directory
 * @returns {boolean}
 */
export function isInWorkspace(filePath, workspace) {
  try {
    const resolvedWorkspace = _normForCompare(fs.realpathSync(workspace));
    let resolvedPath;
    try {
      resolvedPath = _normForCompare(fs.realpathSync(filePath));
    } catch {
      // File doesn't exist yet — walk up to nearest existing ancestor
      resolvedPath = _normForCompare(resolveViaAncestor(filePath));
    }
    return resolvedPath === resolvedWorkspace ||
      resolvedPath.startsWith(resolvedWorkspace + path.sep);
  } catch {
    // Workspace itself doesn't exist — fall back to strict path.resolve
    const resolvedPath = _normForCompare(filePath);
    const resolvedWorkspace = _normForCompare(workspace);
    return resolvedPath === resolvedWorkspace ||
      resolvedPath.startsWith(resolvedWorkspace + path.sep);
  }
}

/**
 * Resolve a non-existent path by finding the nearest existing ancestor,
 * resolving it through fs.realpathSync (following symlinks), then appending
 * the remaining path segments. This prevents symlink escape attacks where
 * a symlink inside the workspace points outside it.
 *
 * Example: /workspace/evil-link/file.txt where evil-link -> /etc
 *   -> ancestor = /workspace/evil-link, realpath = /etc
 *   -> result = /etc/file.txt  (correctly detected as outside workspace)
 *
 * Includes circular-symlink detection and MAX_DEPTH guard.
 *
 * @param {string} targetPath - Path to resolve
 * @returns {string} Resolved path
 */
export function resolveViaAncestor(targetPath) {
  const absolute = path.resolve(targetPath);
  let current = absolute;
  const trailing = [];
  const visited = new Set();
  const MAX_DEPTH = 40;

  while (current !== path.dirname(current) && trailing.length < MAX_DEPTH) {
    // Circular symlink detection
    if (visited.has(current)) {
      return absolute;
    }
    visited.add(current);
    try {
      const real = fs.realpathSync(current);
      const resolved = trailing.length > 0
        ? path.join(real, ...trailing.reverse())
        : real;
      return resolved;
    } catch {
      trailing.push(path.basename(current));
      current = path.dirname(current);
    }
  }

  // No ancestor could be resolved — return absolute
  return absolute;
}

/**
 * Validate a git ref (loose version for branch names and common refs).
 * Allows hex hashes (7-40 chars), branch-like names (with / and -),
 * and common refs like HEAD, refs/heads/main, etc.
 *
 * @param {string} ref
 * @returns {boolean}
 */
export function isValidGitRef(ref) {
  if (!ref || typeof ref !== "string") return false;
  // Pure hex hash (7-40 chars)
  if (/^[a-fA-F0-9]{7,40}$/.test(ref)) return true;
  // Branch-like names: alphanumeric, hyphens, slashes, dots, underscores
  // Must not start/end with . or contain .. or end with .lock
  if (/^[a-zA-Z0-9][a-zA-Z0-9._\/-]*[a-zA-Z0-9]$/.test(ref) &&
      !ref.includes("..") &&
      !ref.endsWith(".lock")) {
    return true;
  }
  // Special refs
  if (/^(HEAD|FETCH_HEAD|ORIG_HEAD|MERGE_HEAD|CHERRY_PICK_HEAD)$/.test(ref)) return true;
  return false;
}

/**
 * Validate a git hash (strict version - only pure SHA hex).
 * Only allows hexadecimal strings of 7-40 characters.
 *
 * @param {string} ref
 * @returns {boolean}
 */
export function isValidGitHash(ref) {
  if (!ref || typeof ref !== "string") return false;
  return /^[a-fA-F0-9]+$/.test(ref) && ref.length >= 7 && ref.length <= 40;
}

File: rust-port/crates/mercury-safety/src/path_safety.rs
// Mercury Code - Shared Path Safety Utilities
// Unified workspace boundary checks, symlink-safe resolution, and git ref validation.
// Ported from: src/utils/path-safety.js

use std::path::{Path, PathBuf};

/// Normalize a path for comparison: resolve to absolute, normalize separators.
/// On Windows, lowercases the path for case-insensitive comparison.
fn norm_for_compare(p: &Path) -> PathBuf {
    let resolved = if p.is_absolute() {
        p.to_path_buf()
    } else {
        std::env::current_dir()
            .unwrap_or_else(|_| PathBuf::from("/"))
            .join(p)
    };
    // Normalize the path by resolving . and .. components lexically
    let normalized = lexical_normalize(&resolved);
    if cfg!(windows) {
        PathBuf::from(normalized.to_string_lossy().to_lowercase())
    } else {
        normalized
    }
}

/// Lexically normalize a path (resolve `.` and `..` without filesystem access).
fn lexical_normalize(p: &Path) -> PathBuf {
    let mut components = Vec::new();
    for component in p.components() {
        match component {
            std::path::Component::CurDir => {}
            std::path::Component::ParentDir => {
                if let Some(last) = components.last() {
                    let is_root = matches!(last, std::path::Component::RootDir)
                        || matches!(last, std::path::Component::Prefix(_));
                    if !is_root {
                        components.pop();
                        continue;
                    }
                }
                // Keep ParentDir if we can't go up further
            }
            _ => {}
        }
        components.push(component);
    }
    if components.is_empty() {
        PathBuf::from(".")
    } else {
        components.iter().collect()
    }
}

/// Unified workspace boundary check (symlink-safe).
/// Uses `std::fs::canonicalize` to resolve symlinks before comparison.
/// Falls back to lexical normalization if canonicalization fails.
///
/// # Arguments
/// * `file_path` - Path to check
/// * `workspace` - Workspace root directory
///
/// # Returns
/// `true` if `file_path` is within `workspace`
pub fn is_in_workspace(file_path: &Path, workspace: &Path) -> bool {
    // Try to canonicalize both paths (resolves symlinks)
    let resolved_workspace = match std::fs::canonicalize(workspace) {
        Ok(p) => norm_for_compare(&p),
        Err(_) => {
            // Workspace doesn't exist - fall back to lexical normalization
            let resolved_path = norm_for_compare(file_path);
            let resolved_ws = norm_for_compare(workspace);
            return is_path_within(&resolved_path, &resolved_ws);
        }
    };

    let resolved_path = match std::fs::canonicalize(file_path) {
        Ok(p) => norm_for_compare(&p),
        Err(_) => {
            // File doesn't exist yet - resolve via ancestor
            norm_for_compare(&resolve_via_ancestor(file_path))
        }
    };

    is_path_within(&resolved_path, &resolved_workspace)
}

/// Check if `target` is within `root` (path prefix check).
fn is_path_within(target: &Path, root: &Path) -> bool {
    if target == root {
        return true;
    }
    // Use starts_with which properly handles path components
    target.starts_with(root)
}

/// Resolve a non-existent path by finding the nearest existing ancestor,
/// resolving it through `canonicalize` (following symlinks), then appending
/// the remaining path segments.
///
/// This prevents symlink escape attacks where a symlink inside the workspace
/// points outside it.
///
/// Example: `/workspace/evil-link/file.txt` where `evil-link -> /etc`
///   -> ancestor = `/workspace/evil-link`, realpath = `/etc`
///   -> result = `/etc/file.txt` (correctly detected as outside workspace)
pub fn resolve_via_ancestor(target_path: &Path) -> PathBuf {
    let absolute = if target_path.is_absolute() {
        target_path.to_path_buf()
    } else {
        std::env::current_dir()
            .unwrap_or_else(|_| PathBuf::from("/"))
            .join(target_path)
    };

    let mut current = absolute.clone();
    let mut trailing: Vec<std::ffi::OsString> = Vec::new();
    let mut visited = std::collections::HashSet::new();
    const MAX_DEPTH: usize = 40;

    loop {
        if trailing.len() >= MAX_DEPTH {
            return absolute;
        }

        let parent = match current.parent() {
            Some(p) if p != current => p.to_path_buf(),
            _ => return absolute, // Reached root
        };

        // Circular symlink detection
        if !visited.insert(current.clone()) {
            return absolute;
        }

        match std::fs::canonicalize(&current) {
            Ok(real) => {
                if trailing.is_empty() {
                    return real;
                }
                let mut resolved = real;
                for component in trailing.into_iter().rev() {
                    resolved.push(component);
                }
                return resolved;
            }
            Err(_) => {
                if let Some(file_name) = current.file_name() {
                    trailing.push(file_name.to_os_string());
                }
                current = parent;
            }
        }
    }
}

/// Validate a git ref (loose version for branch names and common refs).
/// Allows hex hashes (7-40 chars), branch-like names (with `/` and `-`),
/// and common refs like HEAD, refs/heads/main, etc.
pub fn is_valid_git_ref(git_ref: &str) -> bool {
    if git_ref.is_empty() {
        return false;
    }

    // Pure hex hash (7-40 chars)
    if git_ref.len() >= 7
        && git_ref.len() <= 40
        && git_ref.chars().all(|c| c.is_ascii_hexdigit())
    {
        return true;
    }

    // Special refs
    if matches!(
        git_ref,
        "HEAD" | "FETCH_HEAD" | "ORIG_HEAD" | "MERGE_HEAD" | "CHERRY_PICK_HEAD"
    ) {
        return true;
    }

    // Branch-like names: alphanumeric, hyphens, slashes, dots, underscores
    // Must not start/end with . or contain .. or end with .lock
    if git_ref.len() >= 4
        && git_ref
            .chars()
            .all(|c| c.is_ascii_alphanumeric() || "._/-".contains(c))
        && git_ref
            .chars()
            .next()
            .is_some_and(|c| c.is_ascii_alphanumeric())
        && git_ref
            .chars()
            .last()
            .is_some_and(|c| c.is_ascii_alphanumeric())
        && !git_ref.contains("..")
        && !git_ref.ends_with(".lock")
    {
        return true;
    }

    false
}

/// Validate a git hash (strict version - only pure SHA hex).
/// Only allows hexadecimal strings of 7-40 characters.
pub fn is_valid_git_hash(git_ref: &str) -> bool {
    if git_ref.is_empty() {
        return false;
    }
    git_ref.len() >= 7
        && git_ref.len() <= 40
        && git_ref.chars().all(|c| c.is_ascii_hexdigit())
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;
    use tempfile::TempDir;

    #[test]
    fn test_is_in_workspace_basic() {
        let tmp = TempDir::new().unwrap();
        let ws = tmp.path();
        let file = ws.join("src/main.rs");
        fs::create_dir_all(ws.join("src")).unwrap();
        fs::write(&file, "fn main() {}").unwrap();

        assert!(is_in_workspace(&file, ws));
        assert!(is_in_workspace(ws, ws));
    }

    #[test]
    fn test_is_in_workspace_outside() {
        let tmp = TempDir::new().unwrap();
        let ws = tmp.path().join("project");
        fs::create_dir_all(&ws).unwrap();

        let outside = tmp.path().join("other/file.txt");
        assert!(!is_in_workspace(&outside, &ws));
    }

    #[test]
    fn test_is_in_workspace_traversal() {
        let tmp = TempDir::new().unwrap();
        let ws = tmp.path().join("project");
        fs::create_dir_all(&ws).unwrap();

        // Attempt path traversal
        let traversal = ws.join("../other/file.txt");
        assert!(!is_in_workspace(&traversal, &ws));
    }

    #[test]
    fn test_resolve_via_ancestor_existing() {
        let tmp = TempDir::new().unwrap();
        let existing = tmp.path().join("exists");
        fs::create_dir_all(&existing).unwrap();

        let non_existing = existing.join("new_file.txt");
        let resolved = resolve_via_ancestor(&non_existing);
        assert!(resolved.starts_with(std::fs::canonicalize(&existing).unwrap()));
    }

    #[test]
    fn test_is_valid_git_ref() {
        // Valid hex hashes
        assert!(is_valid_git_ref("abc1234"));
        assert!(is_valid_git_ref("abc1234567890abcdef1234567890abcdef12345"));

        // Special refs
        assert!(is_valid_git_ref("HEAD"));
        assert!(is_valid_git_ref("FETCH_HEAD"));

        // Branch names
        assert!(is_valid_git_ref("main"));
        assert!(is_valid_git_ref("feature/my-branch"));
        assert!(is_valid_git_ref("refs/heads/main"));

        // Invalid
        assert!(!is_valid_git_ref(""));
        assert!(!is_valid_git_ref("abc")); // too short for hash, too short for branch
        assert!(!is_valid_git_ref("..bad"));
        assert!(!is_valid_git_ref("branch.lock"));
    }

    #[test]
    fn test_is_valid_git_hash() {
        assert!(is_valid_git_hash("abc1234"));
        assert!(is_valid_git_hash("abc1234567890abcdef1234567890abcdef12345"));
        assert!(!is_valid_git_hash("abc123")); // too short
        assert!(!is_valid_git_hash("main")); // not hex
        assert!(!is_valid_git_hash("")); // empty
    }

    #[cfg(unix)]
    #[test]
    fn test_symlink_escape_detection() {
        let tmp = TempDir::new().unwrap();
        let ws = tmp.path().join("workspace");
        fs::create_dir_all(&ws).unwrap();

        let target_dir = tmp.path().join("outside");
        fs::create_dir_all(&target_dir).unwrap();

        // Create symlink inside workspace pointing outside
        let link = ws.join("evil-link");
        std::os::unix::fs::symlink(&target_dir, &link).unwrap();

        let file_via_link = link.join("secret.txt");
        // resolve_via_ancestor should resolve through symlink
        let resolved = resolve_via_ancestor(&file_via_link);
        // The resolved path should be under target_dir, not workspace
        assert!(!is_in_workspace(&resolved, &ws));
    }
}
